### ISSUE TITLE
feat: Improve redirects 

### ISSUE THREAD

**Issue by neznaika0** [michalsn/codeigniter-htmx#79]
I continue to work with HTMX.
Should I redefine `RedirectResponse`(`back(), route(), to()`) and update `redirect()`?
```php
$responseHTMX = single_service('redirectresponse')
        ->hxRedirect((string) previous_url(true))
        ->with(...$flashes)
        ;

$responseCI = redirect()->back()->with(...$flashes);
```
When the request came as HTMX, the redirect will have the initial headers. By making a regular redirect, the response will remain as for HTMX. Therefore, the second entry does not make sense.
Now we need to duplicate the code for different cases (HTMX or a regular request). Would it be better to combine these conditions inside the new functions?

**Comment by neznaika0** [comment]
Also, I have not found a function for the relative path. It would be great to simplify `hxLocation()` when passing the full path:

```php
$responseHTMX = single_service('redirectresponse')->hxLocation((string) previous_url(true));
public function hxLocation(...): RedirectResponse {
    // $path = 'http://example.local/admin/users?searchQuery=John&sort=ASC';
    /**
     * @var URI $uri
     */
    $uri  = single_service('uri', $path)->withScheme('')->setHost('');
    $data = ['path' => '/' . ltrim((string) $uri, '/')];
    // to '/admin/users?searchQuery=John&sort=ASC'
}
```

**Comment by michalsn** [comment]
The `redirect()` and `hxRedirect()` behave differently and set different headers.

If we send `htmx` request it doesn't mean automatically we will want to use `hxRedirect()` and vice versa.

I want to keep the current methods intact as they are the equivalent of the pure `htmx` implementation. However, I'm open to introducing some sort of helper function that would serve for auto-recognizing the type of request and then serving different responses.

```php
function hxRedirect()
{
 // it htmx then this
 // and if normal request then that
}
```

**Comment by neznaika0** [comment]
Yes, I mentioned the headers. This is the reason for the change.
It doesn't make sense to have a 30x redirect if the request comes from HTMX. It only stealthily replace content from new route. And it adds complexity if 30x is supposed to reload the page.

Example, from the list of users with the "user-list" fragment, click the Ban/Unban button. With HTMX, you need `hxRedirect() / hxLocation()` and return fragments, but without it (click on the link `<a>`) you need 30x redirects with the full page. Otherwise (now), after the 30x redirect, the page sends a fragment of the "user-list".

Therefore, duplicate conditions arise.
```php
        if ($this->request->isHtmx()) {
            return $responseHTMX->with(...$flashError);
        }

        return redirect()->back()->with(...$flashError);
```
Changing the function would be a good solution to use a single code. `hxRedirect()` may help, but the code for the methods `back()`... also needs to be improved and `hxRedirect()` loses its meaning.

PS:  Sorry for the machine translation, if not everything is clearly described

**Comment by michalsn** [comment]
You can name this helper function `autoRedirect()` or something similar - this can be discussed as we go. 

Unfortunately, I will not approve any changes to the existing methods, because we need to be able to decide whether we use `htmx` response or not. This decision cannot blindly rely on the `HX-Request` header by default.

**Comment by neznaika0** [comment]
I heard you. Maybe someday there will be users with a similar problem and we will come back to this. It's easier for me to modify my project

**Comment by neznaika0** [comment]
Another: Is the body really needed when reloading?
With large response output, it looks expensive and useless
```php
return $this->setStatusCode(200)->setHeader('HX-Location', json_encode($data))->setBody('');
return $this->setStatusCode(200)->setHeader('HX-Redirect', $uri)->setBody('');
```

**Comment by neznaika0** [comment]
https://github.com/michalsn/codeigniter-htmx/issues/79#issuecomment-2410660298
Real improve?

**Comment by michalsn** [comment]
> [#79 (comment)](https://github.com/michalsn/codeigniter-htmx/issues/79#issuecomment-2410660298) 
Real improve?

Yes, but I would implement it a bit differently:

```php
public function hxLocation(...): RedirectResponse {

    if (str_starts_with($path, 'http')) {
        $path = (string) service('uri', $path, false)->withScheme('')->setHost('');
    }

    $data = ['path' => '/' . ltrim($path, '/')];
    
    // ...
}
```

Plus this would need a mention in the docs.

**Comment by neznaika0** [comment]
There may be a problem with http verification if the URL is "http/users/123"

**Comment by michalsn** [comment]
Good point, you can change it to `http://` and `https://`

**Comment by neznaika0** [comment]
I won't be able to create a PR for the next few weeks.

 I would not use a shared service, but get a new. There may be a conflict when the URI is used outside the redirect
